### PR TITLE
Vite: Import preview runtime as ordinary module

### LINKED ISSUE #777: Basic Usage example from documentation not working
### Environment

- Operating System: Linux
- Node Version:     v18.20.3
- Nuxt Version:     3.13.1
- CLI Version:      3.13.2
- Nitro Version:    2.9.7
- Package Manager:  npm@10.2.3
- Builder:          -
- User Config:      devtools, modules, compatibilityDate
- Runtime Modules:  @nuxtjs/storybook@8.2.0
- Build Modules:    -

### Reproduction

https://stackblitz.com/github/nuxt-modules/storybook/tree/main/examples/starter?file=components%2FMyNuxtWelcome.stories.js&embed=1

### Describe the bug

I've tried to set up this module for the last 2 days on a blank project and it simply will not start. Been running into various issues (several reported here), but can't seem to crack it and actually get any of the stories to render. Then I realised even the examples in the documentation are not actually starting up (just keep loading): https://storybook.nuxtjs.org/examples/basic
image

There seem to be some dependency issues as per the terminal
image

I've been trying various different combinations of the dependencies but they all lead to different issues. I'd really appreciate if someone could just post a recent working example so I could just fork off of that.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by tobiasdiez**
Closes https://github.com/nuxt-modules/storybook/issues/777.

## What I did

Currently, the preview runtime is manually exposed in the builder (in dev mode via `express.static` and in build mode via maual copying).

This has several problems:
- Makes the whole setup complex and thus suceptable to breaking (as demonstrated by https://github.com/nuxt-modules/storybook/issues/777 where the preview runtime cannot be found for some reason in dev mode)
- The preview runtime has to be shipped in a minified version because it would end up on the server in exactly this version
- But it may still contain too much code since not all if it may be used in a given storybook.

This PR improves the situation by just importing the preview runtime as an ordinary modue (for the vite builder). So it goes through the vite pipeline just as every other module. As a nice side-effect, express is now no longer needed for the vite builder.

#### The changes in this PR are covered in the following automated tests:
- [ ] stories
- [ ] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

_This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!_

### Documentation

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

| name | before | after | diff | z | % |
| ---- | ------ | ----- | ---- | - | - |
| createSize |  0 B | 0 B | 0 B | - | - |
| generateSize |  77.7 MB | 77.7 MB | 0 B | **1.32** | 0% |
| initSize |  130 MB | 130 MB | -33.2 kB | 0.59 | 0% |
| diffSize |  52.5 MB | 52.4 MB | -33.2 kB | 0.45 | -0.1% |
| buildSize |  6.83 MB | 6.75 MB | -82.2 kB | **-685.76** | -1.2% |
| buildSbAddonsSize |  1.51 MB | 1.51 MB | 0 B | -0.23 | 0% |
| buildSbCommonSize |  195 kB | 195 kB | 0 B | - | 0% |
| buildSbManagerSize |  1.86 MB | 1.86 MB | 0 B | 0.03 | 0% |
| buildSbPreviewSize |  271 kB | 0 B | -271 kB | **-54020.65** | 🔰-Infinity% |
| buildStaticSize |  0 B | 0 B | 0 B | - | - |
| buildPrebuildSize |  3.84 MB | 3.56 MB | -271 kB | **-2259.5** | 🔰-7.6% |
| buildPreviewSize |  3 MB | 3.19 MB | 189 kB | **61805.81** | 🔺5.9% |
| testBuildSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbAddonsSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbCommonSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbManagerSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbPreviewSize |  0 B | 0 B | 0 B | - | - |
| testBuildStaticSize |  0 B | 0 B | 0 B | - | - |
| testBuildPrebuildSize |  0 B | 0 B | 0 B | - | - |
| testBuildPreviewSize |  0 B | 0 B | 0 B | - | - |

| name | before | after | diff | z | % |
| ---- | ------ | ----- | ---- | - | - |
| createTime |  7.2s | 8.1s | 886ms | -0.82 | 10.8% |
| generateTime |  20s | 23.3s | 3.3s | 0.57 | 14.2% |
| initTime |  12.7s | 15.5s | 2.8s | 0.56 | 18.2% |
| buildTime |  7.9s | 8.8s | 881ms | -0.19 | 10% |
| testBuildTime |  0ms | 0ms | 0ms | - | - |
| devPreviewResponsive |  5.1s | 4.7s | -395ms | -0.35 | -8.3% |
| devManagerResponsive |  3.6s | 3.6s | -13ms | 0.2 | -0.4% |
| devManagerHeaderVisible |  633ms | 561ms | -72ms | -0.3 | -12.8% |
| devManagerIndexVisible |  693ms | 588ms | -105ms | -0.8 | -17.9% |
| devStoryVisibleUncached |  1.7s | 1.9s | 188ms | 1.07 | 9.5% |
| devStoryVisible |  692ms | 587ms | -105ms | -0.59 | -17.9% |
| devAutodocsVisible |  561ms | 635ms | 74ms | 1.23 | 11.7% |
| devMDXVisible |  511ms | 616ms | 105ms | 0.95 | 17% |
| buildManagerHeaderVisible |  520ms | 601ms | 81ms | 0.18 | 13.5% |
| buildManagerIndexVisible |  532ms | 674ms | 142ms | 0.57 | 21.1% |
| buildStoryVisible |  514ms | 552ms | 38ms | -0.22 | 6.9% |
| buildAutodocsVisible |  458ms | 459ms | 1ms | -0.21 | 0.2% |
| buildMDXVisible |  425ms | 460ms | 35ms | -0.12 | 7.6% |

## Greptile Summary

This pull request simplifies the handling of the preview runtime in the Vite builder for Storybook. Key changes include:

- Imports preview runtime as an ordinary module instead of manual exposure
- Removes 'express' dependency from '@storybook/builder-vite' package
- Modifies Vite build configuration to remove './sb-preview/runtime.js' from external rollup options
- Updates HMR handling for better Vite compatibility
- Removes manual handling of preview runtime via express.static and file copying
- Adds new exports for preview runtime and manager globals runtime in core package
- Updates Vue component meta plugin to modify file exclusion patterns

These changes aim to streamline the setup, reduce complexity, and improve integration with Vite's build process.

### CODE REVIEW

**Comment by ndelangen** [comment]
I suspect we likely won't want to do this for 2 reasons:

The first being:
 
buildTime | 8.9s | 11.3s | 2.3s | 1.36 | 🔺21.1%
-- | -- | -- | -- | -- | --


We prebundle, to minimize build-time, and optimize start-up time.
We know this causes a bit of a size increase, and it's likely not the most intuitive.. but in this case, we deem the trade-off worth it.

And the second being how this would make the vite builder work very differently from how other builders work, which will cause unexpected maintenance problems.

**Comment by tobiasdiez** [comment]
> I suspect we likely won't want to do this for 2 reasons:
> 
> The first being:
> buildTime 	8.9s 	11.3s 	2.3s 	1.36 	🔺21.1%
> 

Did you confirm this locally? The benchmark results are varying very much for my PRs, so I don't know how reliable these numbers are. My expectation would be that there is indeed a very small increase in the cold startup time (because the very first time vite has to process one more file) but afterwards there is a small improvement in the warm startup time (because vite can optimize the import, cache it etc).

> And the second being how this would make the vite builder work very differently from how other builders work, which will cause unexpected maintenance problems.

Webpack would very much profit from the same optimization. I just don't have the knowledge to do this.

**Comment by tobiasdiez** [comment]
Thanks @kasperpeulen for looking into this.

If I remember correctly, the problem was a bit vague in the first problem. The `runtime.js`/`preview.jso` or files loaded by it / related files were simply reported as missing or that they couldn't be loaded. Happened only in dev mode. I never figured out why it occurred, it appeared to me that vite had some troubles analyzing / transpiling all files. But maybe it was also simply an undesired effect of the various hacks we have in place (https://github.com/nuxt-modules/storybook/blob/736d46c89c978b872e4cec027d64b1a044589e83/packages/storybook-addon/src/preset.ts#L373 and https://github.com/nuxt-modules/storybook/blob/736d46c89c978b872e4cec027d64b1a044589e83/packages/storybook-addon/src/preset.ts#L299 being candidates).

I'm not even totally sure the issue is fixed with this PR but in either case it should make the setup simpler and more reliable.

File: code/core/scripts/prep.ts [hunk 1]
```diff
@@ -59,7 +59,7 @@ async function run() {
   console.log(isWatch ? 'Watching...' : 'Bundling...');
 
   const files = measure(generateSourceFiles);
-  const packageJson = measure(() => generatePackageJsonFile(entries));
+  const packageJson = measure(() => generatePackageJsonFile(entries.concat(bundles)));
```
Comment (kasperpeulen) [inline code/core/scripts/prep.ts:62]: I can not really place this change.

File: code/lib/cli/package.json [hunk 2]
```diff
@@ -187,6 +187,12 @@
       "types": "./core/babel/index.d.ts",
       "import": "./core/babel/index.js",
       "require": "./core/babel/index.cjs"
+    },
+    "./internal/manager/globals-runtime": {
+      "import": "./core/manager/globals-runtime.js"
+    },
```
Comment (kasperpeulen) [inline code/lib/cli/package.json:193]: Where is this one used?

File: code/core/scripts/prep.ts (hunk 1 above)
Comment (tobiasdiez) [reply code/core/scripts/prep.ts:62]: Sorry, it's too long ago. I don't remember why I changed this.

File: code/lib/cli/package.json (hunk 2 above)
Comment (tobiasdiez) [reply code/lib/cli/package.json:193]: It's not used, and only there for symmetry with the preview runtime.

File: code/core/src/preview/runtime.ts [hunk 3]
```diff
@@ -6,25 +6,30 @@ import { globalPackages, globalsNameReferenceMap } from './globals/globals';
...
+  });
+  global.addEventListener('unhandledrejection', ({ reason }: any) => {
+    if (reason.fromStorybook) {
+      global.sendTelemetryError(reason);
+    }
+  });
+}
+
+// TODO: In the future, remove this call to make the module side-effect free
+setup();
```
Comment (kasperpeulen) [inline code/core/src/preview/runtime.ts:35]: Are you not calling this function twice now? Can we not just remove it, now we call it explicitly in codegen-modern-iframe-script.ts?

File: code/core/scripts/prep.ts (hunk 1 above)
Comment (kasperpeulen) [reply code/core/scripts/prep.ts:62]: I see now, it generates the package.json export entries for the bundles.

**Review by kasperpeulen — APPROVED** [review]
LGTM, one comment

File: code/core/src/preview/runtime.ts [hunk 4]
```diff
@@ -6,25 +6,30 @@ import { globalPackages, globalsNameReferenceMap } from './globals/globals';
...
+    if (error.fromStorybook) {
+      global.sendTelemetryError(error);
+    }
+  });
+  global.addEventListener('unhandledrejection', ({ reason }: any) => {
+    if (reason.fromStorybook) {
+      global.sendTelemetryError(reason);
+    }
+  });
+}
```
Comment (kasperpeulen) [inline code/core/src/preview/runtime.ts:32]: Especially those 2 seem problematic, when called twice.

File: code/core/src/preview/runtime.ts (hunk 3 above)
Comment (kasperpeulen) [reply code/core/src/preview/runtime.ts:35]: Okay, I will remove this call, let me know if you have objections

Comment (kasperpeulen) [reply code/core/src/preview/runtime.ts:35]: I see, we can not remove it because of webpack.

File: code/core/src/preview/runtime.ts (hunk 4 above)
Comment (kasperpeulen) [reply code/core/src/preview/runtime.ts]: Made sure we only call the listeners once

**Comment by tobiasdiez** [comment]
@kasperpeulen Thanks for the review!

**Pull request merged**